programs/stake/src/stake_state.rs

```diff
@@ -687,6 +705,21 @@ impl Stake {
             Ok(())
         }
     }
+
+    fn reduce_overage(&mut self, balance: u64, rent_exempt_balance: u64) -> Option<(u64, u64)> {
+        // at this moment, balance >>> rent_excempt_balance (no overflow risk)
+        //   => no there could be possibility of non-rent-exempt stake accounts
+        // this will overwrite innocent deactivating and immeditealy withdrawn stake accounts as well is this ok?
```

Comment:
I think the negative effect here is that the deactivating stake will potentially take longer to fully deactivate, correct?
That seems not terrible to me. The alternative (not correcting deactivating accounts) seems riskier to me.

Reply:
yeah, I agree with rewriting those innocent accounts at once for all for reduced risk.

The exact repercussion is that stake account's user-visible `delegated stake` will suddenly reduced. I hope this doesn't cause any unneeded confusion. I mean, it should affect the sudden reduction of `delegated amount` in CLI and explorer, staking-enabled wallets for stake accounts which aren't affected by bug but rewritten by this.

> deactivating stake will potentially take longer to fully deactivate

Well, this is not true. Yeah, staking calc is very difficult to grasp only in mind. The conclusion is that the rate and overall needed duration (in epoch) of deactivation shouldn't be affected.

I wrote a simple thought experiment like this:

```
A user created 100 SOL stake account some epochs ago and warmed up 40 SOL (= effective stake) up to N-1 epoch 
Then the user deactivates and withdraws 60 SOL at N-1 epoch.
For simplicity, assume no other deactivating cluster stake, no rent exempt calculation

a: if the overage rewrite is run at the start of N epoch:
a.1 `delegation.stake` is reduced to 40 SOL.
a.2 recalculated effective stake would be 16 SOL at N-1 epoch (because delegation.stake is reduced by 2.5x from 100SOL to 40SOL)
a.3 new effective (= now deactivating) stake would be 12 SOL at N epoch (= 16 - (16 * 25%))
a.4 next effective (= deactivating) stake would be 9 SOL at N+1 epoch (= 12 - (12 * 25%))

b: if the overage rewrite isn't run at the start of N epoch:
b.1 `delegation.stake` is remained to be 100 SOL.
b.2 recalculated effective stake would remain 40SOL at N-1 epoch.
b.3 new effective (= now deactivating) stake would be 30 SOL at N epoch (= 40 - (40 * 25%)).
b.4 next effective (= deactivating) stake would be 22.5 SOL at N+1 epoch (= 30 - (30 * 25%))
```